packages/router/src/grpc/offscreen-client.ts

```diff
@@ -0,0 +1,47 @@
+import { WitnessAndBuildRequest } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/view/v1alpha1/view_pb';
+import { WitnessData } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/transaction/v1alpha1/transaction_pb';
+import { InternalRequest, InternalResponse } from '@penumbra-zone/types/src/internal-msg/shared';
+import {
+  ActionBuildMessage,
+  OffscreenMessage,
+} from '@penumbra-zone/types/src/internal-msg/offscreen-types';
+
+const OFFSCREEN_DOCUMENT_PATH = '/offscreen.html';
+
+export const offscreenClient = {
+  buildAction: (
+    arg: WitnessAndBuildRequest,
+    witness: WitnessData,
+    fullViewingKey: string,
+  ): Promise<InternalResponse<ActionBuildMessage>> =>
+    sendOffscreenMessage<ActionBuildMessage>({
+      type: 'BUILD_ACTION',
+      request: {
+        transactionPlan: arg.transactionPlan!.toJson(),
+        witness: witness.toJson(),
+        fullViewingKey,
+        length: arg.transactionPlan!.actions.length,
+      },
+    }),
+};
+
+export const sendOffscreenMessage = async <T extends OffscreenMessage>(
+  req: InternalRequest<T>,
+): Promise<InternalResponse<ActionBuildMessage>> => {
+  await chrome.offscreen.createDocument({
```

Comment:
it also occurs to me that crosstalk between two tabs building transactions could occur. will have some more notes on this in a moment

Reply:
well, maybe that's wrong, it is an awaited response

Reply:
what was wrong with the former detection [method](https://github.com/penumbra-zone/web/pull/318/commits/c29f8f6783bca29645dcff5d87ebc13542eb3a93#diff-da83cef549a23dfcccff524ac5206ac90d96d058a42a7183bd22a59d9451aefdR76)?

I was referencing this [example](https://github.com/GoogleChrome/chrome-extensions-samples/blob/main/functional-samples/cookbook.offscreen-dom/background.js).

Reply:
interesting. the examples seem inconsistent. the doc suggests this method, and one of the examples does use it

https://github.com/GoogleChrome/chrome-extensions-samples/blob/578c684d3e27325d513f924bf079705c5f7a767d/functional-samples/sample.tabcapture-recorder/service-worker.js#L16-L21

but, in mentioning the `getContexts` api, the doc links to your example, and i do see one other example also using the method you identified.

i'm inclined to prefer the `chrome.runtime.getContexts` api, instead of using the browser's `Client` api, as the documentation suggests. also, using the chrome api would avoid declaring a global or casting the global object, which feels incorrect.

Reply:
It properly detected and threw an error when multiple offscreen documents were attempted to open.

Reply:
`chrome.runtime.getContexts` was only added in Chrome 116 https://developer.chrome.com/docs/extensions/reference/api/offscreen

Reply:
> i'm inclined to prefer the `chrome.runtime.getContexts` api,

chrome.runtime.getContexts was only added in Chrome 116 https://developer.chrome.com/docs/extensions/reference/api/offscreen

Reply:
i think it is okay to target latest chrome. iirc we are already using other features from 118 for things like popup lifecycle

Reply:
It's strangely not recognizing `chrome.runtime.getContexts`. @turbocrime can you implement the detection system you're satisfied with?

Reply:
lol. we're using a third-party types package which doesn't provide it

https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/66593

and google's official chrome-types also fails to provide it

https://github.com/search?q=repo%3AGoogleChrome%2Fchrome-types+Context&type=code